src/messaging/ExchangeContext.cpp

```diff
@@ -308,25 +370,80 @@ CHIP_ERROR ExchangeContext::HandleMessage(const PacketHeader & packetHeader, con
     // layer has completed its work on the ExchangeContext.
     Retain();
 
+    messageId   = packetHeader.GetMessageId();
     protocolId  = payloadHeader.GetProtocolID();
     messageType = payloadHeader.GetMessageType();
 
-    // Since we got the response, cancel the response timer.
-    CancelResponseTimer();
+    if (payloadHeader.IsAckMsg())
+    {
+        Optional<uint32_t> ackId = payloadHeader.GetAckId();
 
-    // If the context was expecting a response to a previously sent message, this message
-    // is implicitly that response.
-    SetResponseExpected(false);
+        if (ackId.HasValue())
+        {
+            err = mReliableMessageContext.HandleRcvdAck(ackId.Value());
+        }
+        else
+        {
+            err = CHIP_ERROR_DECODE_FAILED;
```

Comment:
Can this actually happen?  I mean, it can if `PayloadHeader::Decode` failed, but I'd expect we would not get here if that happened.  I would rather we just removed the dead code and wrote this as:
```
if (payloadHeader.IsAckMsg())) 
{
  err = mReliableMessageContext.HandleRcvdAck(payloadHeader.GetAckId().Value());
}
```

The fact that we are storing the same boolean state in two places (the `Optional` and the flag byte) is not great.  I filed https://github.com/project-chip/connectedhomeip/issues/4083 on this.

Reply:
Acked